internal/engine/dockerprune/docker_pruner.go

```diff
@@ -0,0 +1,176 @@
+	dp.disabledForTesting = true
+}
+
+func (dp *DockerPruner) OnChange(ctx context.Context, st store.RStore) {
+	if dp.disabledForTesting {
+		return
+	}
+
+	if dp.started {
+		return
+	}
+
+	state := st.RLockState()
+	defer st.RUnlockState()
+
+	if state.DockerPruneSettings.Disabled {
+		return
+	}
+
+	if err := dp.sufficientVersionError(); err != nil {
+		return
+	}
+
+	// wait until state has been kinda initialized, and there's at least one Docker build
+	if !state.TiltStartTime.IsZero() && state.LastTiltfileError() == nil && state.HasDockerBuild() {
+		dp.started = true
+		go func() {
+			select {
+			case <-time.After(time.Minute * 2):
+				dp.Prune(ctx, defaultMaxAge) // report once pretty soon after startup...
+			case <-ctx.Done():
+				return
+			}
+
+			for {
+				select {
+				case <-time.After(dockerPruneInterval):
+					// and once every <interval> thereafter
+					dp.Prune(ctx, defaultMaxAge)
+				case <-ctx.Done():
+					return
+				}
+			}
+		}()
+	}
+}
+
+func (dp *DockerPruner) Prune(ctx context.Context, maxAge time.Duration) {
+	// For future: dispatch event with output/errors to be recorded
+	//   in engineState.TiltSystemState on store (analogous to TiltfileState)
+	err := dp.prune(ctx, maxAge)
+	if err != nil {
+		logger.Get(ctx).Infof("[Docker Prune] error running docker prune: %v", err)
+	}
+}
+
+func (dp *DockerPruner) prune(ctx context.Context, maxAge time.Duration) error {
+	l := logger.Get(ctx)
+	if err := dp.sufficientVersionError(); err != nil {
+		l.Debugf("[Docker Prune] skipping Docker prune, Docker API version too low:\t%v", err)
+		return nil
+	}
+
+	f := filters.NewArgs(
+		filters.Arg("label", docker.BuiltByTiltLabelStr),
+		filters.Arg("until", maxAge.String()),
+	)
+
+	opts := types.BuildCachePruneOptions{
+		All:         true,
+		KeepStorage: 524288000, // 500MB -- TODO: make configurable
+		Filters:     f,
+	}
+	cacheReport, err := dp.dCli.BuildCachePrune(ctx, opts)
+	if err != nil {
+		if !strings.Contains(err.Error(), `"build prune" requires API version`) {
+			return err
+		}
+		l.Debugf("[Docker Prune] skipping build cache prune, Docker API version too low:\t%s", err)
+	} else {
+		prettyPrintCachePruneReport(cacheReport, l)
+	}
+
+	containerReport, err := dp.dCli.ContainersPrune(ctx, f)
+	if err != nil {
+		return err
+	}
+	prettyPrintContainersPruneReport(containerReport, l)
+
+	f.Add("dangling", "0") // prune all images, not just dangling ones
+	imgReport, err := dp.dCli.ImagesPrune(ctx, f)
+	if err != nil {
+		return err
+	}
+	prettyPrintImagesPruneReport(imgReport, l)
+
+	return nil
+}
+
+func (dp *DockerPruner) sufficientVersionError() error {
+	return dp.dCli.NewVersionError("1.30", "image | container prune with filter: label")
+}
+
+func prettyPrintCachePruneReport(report *types.BuildCachePruneReport, l logger.Logger) {
+	// TODO: human-readable space reclaimed
+	l.Infof("[Docker Prune] removed %d caches, reclaimed %d bytes", len(report.CachesDeleted), report.SpaceReclaimed)
```

Comment:
I wonder if these will be useful or distracting in practice, and if we should skip logging if SpaceReclaimed is 0. I guess we'll see!

Reply:
Good calling, will have them not output in that case.